File: python/problems/leetcode/0014_longest_common_prefix/solution.py
"""LeetCode 14: Longest Common Prefix."""

from __future__ import annotations


class Solution:
    def longestCommonPrefix(self, strs: list[str]) -> str:
        if not strs:
            return ""
        base = strs[0]
        for i, ch in enumerate(base):
            for s in strs[1:]:
                if i >= len(s) or s[i] != ch:
                    return base[:i]
        return base


if __name__ == "__main__":
    assert Solution().longestCommonPrefix(["flower", "flow", "flight"]) == "fl"
    assert Solution().longestCommonPrefix(["dog", "racecar", "car"]) == ""
    assert Solution().longestCommonPrefix([""]) == ""
    print("0014 OK")

File: cpp/problems/leetcode/0014_longest_common_prefix/solution.cpp
// LeetCode 14
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.empty()) return "";
        const string& base = strs[0];
        for (size_t i = 0; i < base.size(); ++i) {
            char ch = base[i];
            for (size_t k = 1; k < strs.size(); ++k) {
                if (i >= strs[k].size() || strs[k][i] != ch)
                    return base.substr(0, i);
            }
        }
        return base;
    }
};

int main() {
    vector<string> a{"flower", "flow", "flight"};
    assert(Solution().longestCommonPrefix(a) == "fl");
    vector<string> b{"dog", "racecar", "car"};
    assert(Solution().longestCommonPrefix(b) == "");
    vector<string> c{""};
    assert(Solution().longestCommonPrefix(c) == "");
    cout << "0014 OK" << endl;
    return 0;
}
